src/core/basetypes/MCDataStreamDecoder.cpp

```diff
@@ -0,0 +1,286 @@
+#include "MCDataStreamDecoder.h"
+
+#include "MCString.h"
+#include "MCUtils.h"
+
+#include <libetpan/libetpan.h>
+
+using namespace mailcore;
+
+DataStreamDecoder::DataStreamDecoder()
+{
+    mFilename = NULL;
+    mEncoding = Encoding7Bit;
+    mDecodedData = NULL;
+    mRemainingData = NULL;
+    mRewriteFile = true;
+}
+
+DataStreamDecoder::~DataStreamDecoder()
+{
+    MC_SAFE_RELEASE(mRemainingData);
+    MC_SAFE_RELEASE(mDecodedData);
+    MC_SAFE_RELEASE(mFilename);
+}
+
+void DataStreamDecoder::setEncoding(Encoding encoding)
+{
+    mEncoding = encoding;
+}
+
+void DataStreamDecoder::setFilename(String * filename)
+{
+    MC_SAFE_REPLACE_COPY(String, mFilename, filename);
+}
+
+Data * DataStreamDecoder::decodedData()
+{
+    return mDecodedData;
+}
+
+static size_t uudecode(const char * text, size_t size, char * dst, size_t dst_buf_size)
+{
+    unsigned int count = 0;
+    const char *b = text;		/* beg */
+    const char *s = b;			/* src */
+    const char *e = b+size;			/* end */
+    char *d = dst;
+    int out = (*s++ & 0x7f) - 0x20;
+
+    /* don't process lines without leading count character */
+    if (out < 0)
+        return size;
+
+    /* dummy check. user must allocate buffer with appropriate length */
+    if (dst_buf_size < out)
+        return size;
+
+    /* don't process begin and end lines */
+    if ((strncasecmp((const char *)b, "begin ", 6) == 0) ||
+        (strncasecmp((const char *)b, "end",    3) == 0))
+        return size;
+
+    //while (s < e - 4)
+    while (s < e && count < out)
+    {
+        int v = 0;
+        int i;
+        for (i = 0; i < 4; i += 1) {
+            char c = *s++;
+            v = v << 6 | ((c - 0x20) & 0x3F);
+        }
+        for (i = 2; i >= 0; i -= 1) {
+            char c = (char) (v & 0xFF);
+            d[i] = c;
+            v = v >> 8;
+        }
+        d += 3;
+        count += 3;
+    }
+    return count;
+}
+
+static void decodedPartDeallocator(char * decoded, unsigned int decoded_length) {
+    mailmime_decoded_part_free(decoded);
+}
+
+static Data * decodeData(Data *encodedData, Encoding encoding, bool partialContent, Data ** pRemainingData)
```

Comment:
Move that function to a new file. The code will be shared with  Data::decodedDataUsingEncoding()